conformance/src/main/scala/za/co/absa/enceladus/conformance/interpreter/DynamicInterpreter.scala

```diff
@@ -72,29 +93,44 @@ object DynamicInterpreter {
           case r: NegationConformanceRule         => NegationRuleInterpreter(r).conform(df)
           case r: CustomConformanceRule           => r.getInterpreter().conform(df)
           case r: MappingConformanceRule          =>
-            if (experimentalMappingRule) {
-              MappingRuleInterpreterNoExplode(r, conformance, explodeContext).conform(df)
+            if (ictx.experimentalMappingRule) {
+              MappingRuleInterpreterGroupExplode(r, ictx.conformance, explodeContext).conform(df)
             } else {
-              MappingRuleInterpreter(r, conformance).conform(df)
+              MappingRuleInterpreter(r, ictx.conformance).conform(df)
             }
+          // Array explode and collect pseudo rules apply array explosions for mapping rule groups that operate on
+          // the same arrays
+          case r: ArrayExplodePseudoRule          =>
+            val (dfOut, ctx) = ExplodeTools.explodeAllArraysInPath(r.outputColumn, df, explodeContext)
+            explodeContext = ctx
+            dfOut
+          case r: ArrayCollectPseudoRule          =>
+            val dfOut = ExplodeTools.revertAllExplosions(df, explodeContext, Some(ErrorMessage.errorColumnName))
+            explodeContext = ExplosionContext()
+            dfOut
           case _ => throw new IllegalStateException(s"Unrecognized rule class: ${rule.getClass.getName}")
         }
-        applyCheckpoint(rule, confd, jobShortName, explodeContext, enableControlFramework)
+        applyRuleCheckpoint(rule, confd, ictx.jobShortName, explodeContext, ictx.enableControlFramework)
     })
+    conformedDf
+  }
 
-    // Imploding all arrays back
-    val implodeDf = ExplodeTools.revertAllExplosions(ds, explodeContext, Some(ErrorMessage.errorColumnName))
-
-    if (enableControlFramework) {
-      implodeDf.setCheckpoint(s"$jobShortName - End", persistInDatabase = false)
-    }
-
-    val explain = ExplainCommand(implodeDf.queryExecution.logical, extended = true)
-    spark.sessionState.executePlan(explain).executedPlan.executeCollect().foreach {
-      r => log.debug("Output Dataset plan: \n" + r.getString(0))
+  private def getConformanceSteps(ictx: InterpreterContext): List[ConformanceRule] = {
```

Comment:
This method and methods associated with it can be moved to the `InterpreterContext` as "getters" of sorts. They are only used for the `InterpreterContext` only anyway.